lib/render/FlowRenderer.cpp

```diff
@@ -322,17 +345,257 @@ FlowRenderer::_paintGL( bool fast )
     glDepthMask(true);
 
     _prepareColormap( params );
-    _renderFromAnAdvection( &_advection, params, fast );
-    /* If the advection is bi-directional */
-    if( _2ndAdvection )
-        _renderFromAnAdvection( _2ndAdvection.get(), params, fast );
+    
+    int ret = 0;
+    
+    if (params->GetValueLong("old_render", 0)) {
+        _renderFromAnAdvectionLegacy( &_advection, params, fast );
+        /* If the advection is bi-directional */
+        if( _2ndAdvection )
+            _renderFromAnAdvectionLegacy( _2ndAdvection.get(), params, fast );
+    } else {
+        // Workaround for how bi-directional was implemented.
+        if (_2ndAdvection)
+            _renderStatus = FlowStatus::SIMPLE_OUTOFDATE;
+        
+        ret |= _renderAdvection(&_advection);
+        /* If the advection is bi-directional */
+        if(_2ndAdvection) {
+            _renderStatus = FlowStatus::SIMPLE_OUTOFDATE;
+            ret |= _renderAdvection(_2ndAdvection.get());
+        }
+    }
+    
     _restoreGLState();
 
+    return ret;
+}
+
+int FlowRenderer::_renderAdvection(const flow::Advection* adv)
+{
+    FlowParams *rp = (FlowParams*)GetActiveParams();
```

Comment:
Please use dynamic_cast for better safety.

Reply:
ok